src/main/kotlin/badasintended/slotlink/block/entity/InterfaceBlockEntity.kt

```diff
@@ -4,57 +4,36 @@ package badasintended.slotlink.block.entity
 
 import badasintended.slotlink.init.BlockEntityTypes
 import badasintended.slotlink.network.ConnectionType
+import badasintended.slotlink.network.ConnectionType.Companion.MASTER
+import badasintended.slotlink.screen.FilterScreenHandler
 import badasintended.slotlink.storage.FilterFlags
 import badasintended.slotlink.storage.FilteredItemStorage
-import java.util.*
-import net.fabricmc.fabric.api.transfer.v1.item.ItemVariant
-import net.fabricmc.fabric.api.transfer.v1.storage.Storage
-import net.fabricmc.fabric.api.transfer.v1.storage.StorageView
+import badasintended.slotlink.storage.FilteredItemStorage.Companion.EMPTY
 import net.fabricmc.fabric.api.transfer.v1.storage.base.CombinedStorage
-import net.fabricmc.fabric.api.transfer.v1.transaction.TransactionContext
 import net.minecraft.block.BlockState
+import net.minecraft.entity.player.PlayerEntity
+import net.minecraft.entity.player.PlayerInventory
+import net.minecraft.screen.ScreenHandlerContext
 import net.minecraft.util.math.BlockPos
 import net.minecraft.util.math.Direction
-import net.minecraft.world.World
 
-class InterfaceBlockEntity(pos: BlockPos, state: BlockState) :
-    ChildBlockEntity(BlockEntityTypes.INTERFACE, ConnectionType.INTERFACE, pos, state),
-    Storage<ItemVariant> {
-
-    private val insertStorageCache = StorageCache(FilterFlags.INSERT)
-    private val extractStorageCache = StorageCache(FilterFlags.EXTRACT)
-
-    override fun insert(resource: ItemVariant, maxAmount: Long, transaction: TransactionContext) =
-        insertStorageCache[transaction].insert(resource, maxAmount, transaction)
-
-    override fun extract(resource: ItemVariant, maxAmount: Long, transaction: TransactionContext) =
-        extractStorageCache[transaction].extract(resource, maxAmount, transaction)
-
-    override fun iterator(transaction: TransactionContext): MutableIterator<StorageView<ItemVariant>> =
-        extractStorageCache[transaction].iterator(transaction)
+private const val flag = FilterFlags.INSERT + FilterFlags.EXTRACT
 
-    private inner class StorageCache(private val flags: Int) {
-
-        private val instances = WeakHashMap<TransactionContext, Storage<ItemVariant>>()
+class InterfaceBlockEntity(pos: BlockPos, state: BlockState) :
+    FilteredBlockEntity(BlockEntityTypes.INTERFACE, ConnectionType.INTERFACE, pos, state) {
+
+    @Suppress("UNUSED_PARAMETER")
+    fun getStorage(unused: Direction): FilteredItemStorage {
+        world?.also { world ->
+            val master = network?.get(MASTER)?.firstOrNull() ?: return EMPTY
+            val storages = master.getStorages(InterfaceBlockEntity::class, world, flag)
+            return FilteredItemStorage(filter, blacklist, flag, CombinedStorage(storages.toList()))
+        }
+        return EMPTY
+    }
 
-        @Synchronized
```

Comment:
Actually this is more than caching. I noticed that *not* returning the same instance of CombinedStorage by transaction caused duplication bugs. However, if you think using weak references is not enough to clean up, we can use the transaction close callback to remove them from the hashmap.

Reply:
Forget this. It might have been caused by the insert/extract dichotomy. With your changes everything seems okay.

Reply:
Instead of caching by transaction context, it instead returns entirely different `Storage` instance when one calls `ItemStorage.SIDED.find`. Should be enough since you're supposed to call that everytime you want to do a transfer.